Add an explanation with diagrams for the free-threaded execution model (#188)
Quansight-Labs/free-threaded-compatibility: docs/porting-extensions.md

development branch. This will ease the work of libraries that depend on yours
to also add support for the free-threaded build.

## Working with the free-threaded CPython interpreter runtime

Many people are surprised to learn that almost all native extensions written

> with the GIL-enabled build in mind build with minimal changes on the

Author: ```suggestion
with the GIL-enabled build in mind compile and run with minimal changes on the
```

free-threaded build. They often ask questions like, "if there is no GIL, doesn't
that mean there's no need to call e.g. `PyGilState_Ensure()` before calling into
the C API and no need to call `Py_BEGIN_ALLOW_THREADS` to release the GIL before

> doing I/O or a long-running computation?". Binding generators like Cython, PyO3,

Author: ```suggestion
doing I/O or a long-running computation?". Bindings generators like Cython, PyO3,
```

scheme, which requires all threads to be detached from the runtime before it can
start. If you don't explicitly detach before doing a long-running operation that
does not require the runtime, the interpreter may be blocked on running the
garbage collector or doing any other operation that requires a globally
consistent view of all threads.

The icons indicating whether threads are attached or detached are still
present. This is because it is still necessary to explicitly attach and detach
from the runtime in the free-threaded build, despite the fact that there isn't a
GIL.

![Attach and detach from the runtime diagram](assets/images/attach-detach-gil.png){ width="400" }
/// caption
Attaching and detaching from the runtime uses the same
code as in the GIL-enabled build
///

As illustrated above, attaching and detaching from the runtime uses exactly the
same code in the free-threaded build as is used in the GIL-enabled build to
acquire and release the GIL. It is an unfortunate naming issue that
`PyGILState_Ensure` and `PyGILState_Release` has "`GIL`" in the name of the
function, despite the lack of a GIL on the free-threaded build. It's likely that

> the C API in future Python version will fix this naming issue.

Author: ```suggestion
the C API in a future Python version will fix this naming issue.
```

Hopefully you now have a better mental model for how native code interacts with
the CPython interpreter runtime in the free-threaded build and how it is similar
to what happens on the GIL-enabled build, and what exactly it means for multiple
threads to simultaneously execute Python code.

You might also see how given that many extension modules were written assuming

> Python can into the extension in one thread at a time, how that might lead to

Author: ```suggestion
Python can call into the extension in one thread at a time, how that might lead to
```

> You might also see how given that many extension modules were written assuming
> Python can into the extension in one thread at a time, how that might lead to
> problematic thread-unsafe behavior if suddenly more than one Python thread can
> simultaneously access any state stored in the extension.

Reviewer: This sentence is somewhat difficult to parse for me. Maybe this is better?

```suggestion
You might also see how writing extension modules assuming that
Python can call into them in one thread at a time might lead to
problematic thread-unsafe behavior, especially now that more than one Python threads can
simultaneously access the extension's state.
```
